viamonvif/device/device.go

```diff
@@ -0,0 +1,326 @@
+// Package device allows communication with an onvif device
+// inspired by https://github.com/use-go/onvif
+package device
+
+import (
+	"bytes"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/beevik/etree"
+	"github.com/viam-modules/viamrtsp/viamonvif/gosoap"
+	"github.com/viam-modules/viamrtsp/viamonvif/xsd/onvif"
+	"go.viam.com/rdk/logging"
+)
+
+const (
+	streamTypeRTPUnicast = "RTP-Unicast"
+	streamSetupProtocol  = "RTSP"
+)
+
+// Xlmns XML Scheam.
+var Xlmns = map[string]string{
+	"onvif":   "http://www.onvif.org/ver10/schema",
+	"tds":     "http://www.onvif.org/ver10/device/wsdl",
+	"trt":     "http://www.onvif.org/ver10/media/wsdl",
+	"tev":     "http://www.onvif.org/ver10/events/wsdl",
+	"tptz":    "http://www.onvif.org/ver20/ptz/wsdl",
+	"timg":    "http://www.onvif.org/ver20/imaging/wsdl",
+	"tan":     "http://www.onvif.org/ver20/analytics/wsdl",
+	"xmime":   "http://www.w3.org/2005/05/xmlmime",
+	"wsnt":    "http://docs.oasis-open.org/wsn/b-2",
+	"xop":     "http://www.w3.org/2004/08/xop/include",
+	"wsa":     "http://www.w3.org/2005/08/addressing",
+	"wstop":   "http://docs.oasis-open.org/wsn/t-1",
+	"wsntw":   "http://docs.oasis-open.org/wsn/bw-2",
+	"wsrf-rw": "http://docs.oasis-open.org/wsrf/rw-2",
+	"wsaw":    "http://www.w3.org/2006/05/addressing/wsdl",
+}
+
+// Device for a new device of onvif and DeviceInfo
+// struct represents an abstract ONVIF device.
+// It contains methods, which helps to communicate with ONVIF device.
+type Device struct {
+	xaddr     *url.URL
+	logger    logging.Logger
+	params    Params
+	endpoints map[string]string
+}
```

Comment:
I don't think we need them at this point - but when this becomes the back end of a discovery model we may want to keep an eye out for adding mutexes for reconfigure-close handling. Just add a note at the top of the file for John/Sean Yu when we make the conversion.

Reply:
how is that?